lang/src/splgen/generator.py

```diff
@@ -0,0 +1,241 @@
+                                for p in f.params.parameters},
+                        mod=set()
+                    ) for f in c.implementation.methods
+                },
+                **{
+                    admin.method.name: Function(
+                        name=admin.method.name,
+                        params={p.name: format_param(p.literal)
+                                for p in admin.method.params.parameters},
+                        is_modifier=True,
+                        mod=set()
+                    ) for admin in model.administration.extension_administrators
+                },
+                manager.method.name: Function(
+                    name=manager.method.name,
+                    params={p.name: format_param(p.literal)
+                            for p in manager.method.params.parameters},
+                    is_modifier=True,
+                    mod=set()
+                )
+            },
+            dependencies=set(),
+        ).resolve_code_and_dependencies(dependencies)
+        .insert_params() for c in model.contract_section.contracts
+    }
+
+    for ca in manager.contract_administrators:
+        for method in ca.methods:
+            contracts[ca.contract.name].functions[method.name].mod.add(
+                manager.method.name)
+
+    for admin in model.administration.extension_administrators:
+        for ca in admin.contract_administrators:
+            for method in ca.methods:
+                contracts[ca.contract.name].functions[method.name].mod.add(
+                    admin.method.name)
+
+    for contract in contracts.values():
+        for fun in contract.functions.values():
+            fun.insert_modifiers()
+
+    with open(output_file, 'w') as f:
+        f.write(f"/* Generated by SPL [{datetime.now()}] */\n\n")
+        f.write("// SPDX-License-Identifier: MIT\npragma solidity ^0.8.17;\n\n")
+        f.write("//"+" DEPENDENCIES ".center(100, '=')+'\n\n')
+        for d in sorted(list(dependencies), key=lambda d: (d.type, d.priority), reverse=True):
+            f.write(d.code+'\n')
+        f.write("//"+" YOUR CONTRACTS ".center(100, '=')+'\n\n')
+        for c in contracts.values():
+            f.write(f"contract {c.name} " + "{\n")
+            for d in sorted(list(c.dependencies), key=lambda d: d.type, reverse=True):
+                f.write('\t'+d.code+'\n')
+            for m in c.functions.values():
+                f.write('\t'+m.code+'\n')
+            f.write("\n}\n")
+
+
+@dataclass
+class Dependency:
+    name: str
+    code: str
+    type: str
+    priority: int = 0
+
+    def __eq__(self, other):
+        if not isinstance(other, Dependency):
+            return False
+        return self.name == other.name and self.code == other.code and self.type == other.type
+
+    def __hash__(self):
+        return hash((self.name, self.code, self.type))
+
+
+@dataclass
+class Function:
+    params: Dict[str, str]
+    name: str
+    mod: Set[str]
```

Comment:
this mod is kinda confusing me tbh.

Reply:
I will change it to something less confusing